test/kubetest/kubernetes.go

```diff
@@ -0,0 +1,332 @@
+				continue
+			}
+			return cond.Status == corev1.ConditionTrue, nil
+		}
+		return false, fmt.Errorf("pod ready condition not found")
+	}
+	return false, nil
+}
+
+func Sleep(d time.Duration) Condition {
+	return func(ctx *ScenarioContext) error {
+		time.Sleep(d)
+		return nil
+	}
+}
+
+type RunOptions struct {
+	ServiceAccount string
+}
+
+func RunSucceeds(client kubernetes.Interface, image string, name string, command []string, opts *RunOptions) Check {
+	return func(ctx *ScenarioContext) error {
+		logs, err := run(client, ctx, image, name, command, opts)
+		if err != nil {
+			_, _ = fmt.Fprint(os.Stderr, string(logs))
+			return err
+		}
+		return nil
+	}
+}
+
+func RunFails(client kubernetes.Interface, image string, name string, command []string, opts *RunOptions) Check {
+	return func(ctx *ScenarioContext) error {
+		logs, err := run(client, ctx, image, name, command, opts)
+		if err == nil {
+			_, _ = fmt.Fprint(os.Stderr, string(logs))
+			return fmt.Errorf("expected run to fail")
+		}
+		if err != errRun {
+			return err
+		}
+		return nil
+	}
+}
+
+var errRun = fmt.Errorf("failed to run")
+
+// run the command and return the Check with the container's logs
+func run(client kubernetes.Interface, ctx *ScenarioContext, image string, name string, command []string, opts *RunOptions) ([]byte, error) {
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: ctx.Namespace,
+			Labels: map[string]string{
+				"app": name,
+			},
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{{
+				Name:    name,
+				Image:   image,
+				Command: command,
+			}},
+			RestartPolicy: corev1.RestartPolicyNever,
+		},
+	}
+
+	if opts != nil {
+		pod.Spec.ServiceAccountName = opts.ServiceAccount
+	}
+
+	ctx.AddFinalizer(func() error {
+		return client.CoreV1().Pods(ctx.Namespace).Delete(pod.ObjectMeta.Name, nil)
+	})
+
+	_, err := client.CoreV1().Pods(ctx.Namespace).Create(pod)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create pod: %v", err)
+	}
+
+	watch, err := client.CoreV1().Pods(ctx.Namespace).Watch(metav1.SingleObject(pod.ObjectMeta))
+	if err != nil {
+		return nil, fmt.Errorf("failed to watch pod: %v", err)
+	}
+
+	for event := range watch.ResultChan() {
+		pod := event.Object.(*corev1.Pod)
+		phase := pod.Status.Phase
+
+		if phase == corev1.PodFailed {
+			logs, _ := podLogs(client, ctx.Namespace, name, name)
+			return logs, errRun
+		}
+		if phase == corev1.PodSucceeded {
+			break
+		}
+	}
+
+	logs, _ := podLogs(client, ctx.Namespace, name, name)
+	return logs, nil
+}
+
+func podLogs(client kubernetes.Interface, namespace, pod, container string) ([]byte, error) {
+	rest := client.CoreV1().Pods(namespace).GetLogs(pod, &corev1.PodLogOptions{
+		Container: container,
+		Follow:    false,
+	})
+
+	stream, err := rest.Stream()
+	if err != nil {
+		return nil, err
+	}
+	return ioutil.ReadAll(stream)
+}
+
+func CreateNamespace(client kubernetes.Interface, name string) error {
```

Comment:
This doesn't need to be public, right?

Reply:
For the future this is something I can see people use to create individual namespaces for each step. Right now it could be private, true.